### PR TITLE
[inductor] Fix a dead buffer removal bug

### PR DESCRIPTION
**Pull request by desertfire**
Summary: For https://github.com/pytorch/torchdynamo/issues/603

### CODE REVIEW

**Comment by desertfire** [comment]
Although my change would fix this problem, but what puzzles me is why this code is generated in the first place (basically we are computing buf725 with uninitialized buf76)?

```
    buf725 = empty_strided((100, 240, 1, 1), (240, 1, 1, 1), device='cuda', dtype=torch.float16)
    # buf76 is removed after fusion
    # buf76 = empty_strided((100, 240, 1, 1), (240, 1, 24000, 24000), device='cuda', dtype=torch.float32)
    kernel28[grid(24000)](buf74, buf76, buf725, 24000, 6400)
```
```
@reduction_heuristics(size_hints=[32768, 8192])
@triton.jit
def kernel28(in_ptr0, in_ptr1, out_ptr1, xnumel: tl.constexpr, rnumel: tl.constexpr, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK, 1])
    xmask = xindex < xnumel
    rbase = tl.reshape(tl.arange(0, RBLOCK), [1, RBLOCK])
    x0 = xindex
    _tmp2 = tl.zeros([XBLOCK, RBLOCK], tl.float32) + 0
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r1 = rindex
        tmp0 = tl.load(in_ptr0 + r1 + (6400*x0), xmask & rmask, eviction_policy='evict_last').to(tl.float32)
        tmp1 = tmp0.to(tl.float32)
        _tmp2 = tl.where(xmask & rmask, _tmp2 + tmp1, _tmp2)
    tmp2 = tl.reshape(tl.sum(_tmp2, 1), [XBLOCK, 1])
    tmp3 = tl.load(in_ptr1 + x0, xmask).to(tl.float32)
    tmp4 = 6400
    tmp5 = tmp2 / tmp4
    tmp6 = tmp5.to(tl.float32)
    tl.store(out_ptr1 + x0, tmp3, xmask)
```

File: torchinductor/scheduler.py [hunk 1]
```diff
@@ -1004,7 +1004,10 @@ def remove_kernel_local_buffers(self):
         # If all the uses of this buffer are also in self.pending_buffer_names,
         # it means the buffer becomes kernel-local after fusion
         for name in self.pending_buffer_names:
-            if name in V.kernel.must_keep_buffers:
+            if (
```
Comment (Chillee) [inline torchinductor/scheduler.py:1007]: I wonder if it's semantically cleaner to just add `input_buffers` into `must_keep_buffers`?

**Review by Chillee — APPROVED** [review]
LGTM

**Comment by Chillee** [comment]
but yeah, that code looks quite strange. Could you post the full log?

**Comment by desertfire** [comment]
> but yeah, that code looks quite strange. Could you post the full log?

The full log is too large. I think this one would be a good candidate for the `minimizer` you were talking about in our group chat. 

You can reproduce the error with `python benchmarks/torchbench.py -d cuda --float16 --inductor --no-skip -k timm_efficientdet` after cherry-picking https://github.com/pytorch/torchdynamo/pull/674.

**Review by jansel — APPROVED** [review]

**Comment by pyjhzwh** [comment]
I checkout out to your branch, and it seems `python benchmarks/torchbench.py -dcuda --inductor-settings --float32 -n1 --inductor --filter mobilenet_v3_large` will give INCORRECT values
```
cuda eval  mobilenet_v3_large                 Similarity score=0.2933821678161621
INCORRECT
ERROR
```
Although on the lasted main branch, there will be `NameError: name 'buf24' is not defined` error

**Comment by pyjhzwh** [comment]
I checkout to you branch, the following command will give incorrect results
```
$python benchmarks/torchbench.py -dcuda --inductor-settings --float32 -n1 --inductor --filter mobilenet_v3_large
cuda eval  mobilenet_v3_large                 Similarity score=0.2933821678161621
INCORRECT
```
Although on main branch, it will failed with `    kernel9[grid(2304)](buf22, buf24, buf160, 2304, 784)
NameError: name 'buf24' is not defined` error

**Pull request merged**